### PR TITLE
haskellPackages.shellFor: improve documentation

### PR DESCRIPTION
**Pull request by cdepillabout**
###### Motivation for this change

I was trying to figure out a problem with benchmarks not being included in `shellFor` (per https://github.com/NixOS/nixpkgs/pull/102323 from @harendra-kumar).

It took me a while to just figure out how `shellFor` works.  I did a small refactoring, and I added some documentation to make it easier to understand.  I think this would be an improvement on the current state of the code.

None of the functionality should be changed.

cc @maralorn and @peti

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by cdepillabout** [comment]
I also have a local commit that adds support for benchmark stuff (which will fix #102323):

```diff
diff --git a/pkgs/development/haskell-modules/make-package-set.nix b/pkgs/development/haskell-modules/make-package-set.nix
index c5604aa4196..8201f0d3aff 100644
--- a/pkgs/development/haskell-modules/make-package-set.nix
+++ b/pkgs/development/haskell-modules/make-package-set.nix
@@ -293,6 +293,10 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
       , # Whether or not to generated a Hoogle database for all the
         # dependencies.
         withHoogle ? false
+      , # Whether or not to include benchmark dependencies of your local
+        # packages.  You should set this to true if you have benchmarks defined
+        # in your local packages that you want to be able to run with cabal benchmark
+        doBenchmark ? false
       , ...
       } @ args:
       let
@@ -398,7 +402,11 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
           version = "0";
           license = null;
         }
-        // packageInputs;
+        // packageInputs
+        // pkgs.lib.optionalAttrs doBenchmark {
+          # `doBenchmark` needs to explicitly be set here because haskellPackages.mkDerivation defaults it to `false`.  If the user wants benchmark dependencies included in their development shell, it has to be explicitly enabled here.
+          doBenchmark = true;
+        };
 
         # This is a pseudo Haskell package derivation that contains all the
         # dependencies for the packages in `selected`.
@@ -418,9 +426,11 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
         # `haskellPackages.mkDerivation`).
         #
         # pkgWithCombinedDepsDevDrv :: Derivation
         pkgWithCombinedDepsDevDrv = pkgWithCombinedDeps.envFunc { inherit withHoogle; };
 
-        mkDerivationArgs = builtins.removeAttrs args [ "packages" "withHoogle" ];
+        mkDerivationArgs = builtins.removeAttrs args [ "packages" "withHoogle" "doBenchmark" ];
 
       in pkgWithCombinedDepsDevDrv.overrideAttrs (old: mkDerivationArgs // {
         nativeBuildInputs = old.nativeBuildInputs ++ mkDerivationArgs.nativeBuildInputs or [];
```

**Comment by cdepillabout** [comment]
@GrahamcOfBorg build tests.haskell-shellFor

**Comment by cdepillabout** [comment]
I get the same hash when building `tests.haskell-shellFor` on both this branch and `haskell-updates`, so that is a strong indicator that I didn't actually change any of the functionality in `shellFor` in this PR.

**Comment by harendra-kumar** [comment]
@cdepillabout my first change was exactly the same change as the one you wrote in https://github.com/NixOS/nixpkgs/pull/102598#issuecomment-720963892 . But then I later changed it to what posted in https://github.com/NixOS/nixpkgs/pull/102323 . I thought that made it simpler. Any reasons for preferring one over the other?

File: pkgs/development/haskell-modules/make-package-set.nix [hunk 1]
```diff
@@ -285,38 +286,143 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
     #
     #     bash$ nix-shell --run "cabal new-build all"
     #     bash$ nix-shell --run "python"
-    shellFor = { packages, withHoogle ? false, ... } @ args:
+    shellFor =
+      { # Packages to create this development shell for.  These are usually
+        # your local packages.
+        packages
+      , # Whether or not to generated a Hoogle database for all the
```
Comment (maralorn) [inline pkgs/development/haskell-modules/make-package-set.nix:293]:
```suggestion
      , # Whether or not to generate a Hoogle database for all the
```

File: pkgs/development/haskell-modules/make-package-set.nix [hunk 2]
```diff
@@ -285,38 +286,143 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
...
-
-            genericBuilderArgs = {
-              inherit pname;
-              version = "0";
-              license = null;
-            } // packageInputs;
-
-          in self.mkDerivation genericBuilderArgs;
+        # A list of the packages we want to build a development shell for.
+        #
```
Comment (maralorn) [inline pkgs/development/haskell-modules/make-package-set.nix:300]:
```suggestion
```

File: pkgs/development/haskell-modules/make-package-set.nix [hunk 3]
```diff
@@ -285,38 +286,143 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
...
+        # This is a list of Haskell package derivations.
+        selected = packages self;
+
+        # This is a list of attribute sets, where the each attribute set
```
Comment (maralorn) [inline pkgs/development/haskell-modules/make-package-set.nix:304]:
```suggestion
        # This is a list of attribute sets, where each attribute set
```

File: pkgs/development/haskell-modules/make-package-set.nix [hunk 4]
```diff
@@ -285,38 +286,143 @@ in package-set { inherit pkgs stdenv callPackage; } self // {
...
+        # corresponds to the build inputs one of the packages input to shellFor.
```
Comment (maralorn) [inline pkgs/development/haskell-modules/make-package-set.nix:305]:
```suggestion
        # corresponds to the build inputs of one of the packages input to shellFor.
```

**Review by maralorn — CHANGES_REQUESTED** [review]
I am too tired right now, too fully grasp every of your comments. But it seems nice.

**Comment by cdepillabout** [comment]
Thanks for the review!

-------------------------------------

@harendra-kumar

> Any reasons for preferring one over the other?

The only reason is that I wanted to explicitly document it.  

Also, those `args` are passed to `envFrom`, which is just a `stdenv.mkDerivation`, not `haskellPackages.mkDerivation`.  `stdenv.mkDerivation` doesn't know anything about `doBenchmark` (that's a `haskellPackages.mkDerivation thing).

**Pull request merged**